File: frontend-pjt/src/api/glossary.js
import { cachedQuery } from './cache'
import { apiGet } from './client'

export const GLOSSARY_PAGE_SIZE = 20

export function fetchGlossaryTerms(params = {}) {
  const requestParams = {
    ...params,
    page: params.page || 1,
    size: params.size || GLOSSARY_PAGE_SIZE,
  }
  const cacheKey = `glossary:list:${JSON.stringify(requestParams)}`

  return cachedQuery(cacheKey, async () => {
    const payload = await apiGet('/glossary', { params: requestParams, raw: true })
    return normalizePaginatedResponse(payload, normalizeTerm)
  })
}

export function fetchGlossaryCategories() {
  return cachedQuery('glossary:categories', async () => {
    const data = await apiGet('/glossary/categories')
    return getItems(data).map(normalizeCategory).filter((item) => item.name)
  })
}

function getItems(data) {
  if (Array.isArray(data)) return data
  if (Array.isArray(data?.data)) return data.data
  if (Array.isArray(data?.items)) return data.items
  if (Array.isArray(data?.results)) return data.results
  if (Array.isArray(data?.terms)) return data.terms
  return []
}

function normalizePaginatedResponse(payload, normalizer) {
  const items = getItems(payload)
  const page = payload?.page || {}

  return {
    items: items.map(normalizer),
    page: {
      number: page.number || 1,
      size: page.size || items.length,
      totalElements: page.total_elements ?? items.length,
      totalPages: page.total_pages || 1,
    },
  }
}

function normalizeCategory(item) {
  return {
    id: item.category_id || item.id || '',
    name: item.category_name || item.category || '',
  }
}

function normalizeTerm(item) {
  const category = item.category || {}

  return {
    id: item.term_id || item.id || item.term_name || item.term || '',
    term: item.term_name || item.term || '',
    categoryId: category.category_id || item.category_id || '',
    category: category.category_name || item.category_name || '',
    level: item.difficulty || item.level || '',
    desc: item.description || item.desc || '',
    example: item.example_text || item.example || '',
  }
}

File: backend-pjt/apps/search/glossary.py
from math import ceil
from django.conf import settings
from apps.search.clients import get_elasticsearch_client

DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100

class SearchUnavailable(Exception):
    pass

def search_glossary(params):
    page = positive_int(params.get("page"), 1)
    size = min(positive_int(params.get("size"), DEFAULT_PAGE_SIZE), MAX_PAGE_SIZE)
    client = get_elasticsearch_client()

    if client is None:
        raise SearchUnavailable("Elasticsearch client is not installed.")

    body = build_glossary_search_body(params, page, size)
    response = client.search(index=settings.ELASTICSEARCH_GLOSSARY_INDEX, **body)
    total = response["hits"]["total"]
    total_value = total.get("value", 0) if isinstance(total, dict) else total

    return {
        "data": [normalize_glossary_hit(hit["_source"]) for hit in response["hits"]["hits"]],
        "page": {
            "number": page,
            "size": size,
            "total_elements": total_value,
            "total_pages": ceil(total_value / size) if total_value else 0,
        },
    }

def build_glossary_search_body(params, page, size):
    must = []
    filters = []
    keyword = (params.get("keyword") or params.get("q") or "").strip()

    if keyword:
        must.append(
            {
                "bool": {
                    "should": [
                        {
                            "term": {
                                "term_name.keyword": {
                                    "value": keyword,
                                    "boost": 20
                                }
                            }
                        },
                        {
                            "multi_match": {
                                "query": keyword,
                                "fields": [
                                    "term_name^5",
                                    "description^2",
                                    "example_text^1"
                                ],
                                "type": "best_fields"
                            }
                        }
                    ],
                    "minimum_should_match": 1
                }
            }
        )

    # Apply category filter
    category_id = params.get("category_id")
    if category_id:
        filters.append({"term": {"category.category_id": category_id}})

    # Apply difficulty filter
    difficulty = params.get("difficulty")
    if difficulty:
        filters.append({"term": {"difficulty": difficulty}})

    query = {"bool": {"filter": filters}}
    if must:
        query["bool"]["must"] = must
    else:
        query["bool"]["must"] = [{"match_all": {}}]

    return {
        "from_": (page - 1) * size,
        "size": size,
        "query": query,
        "sort": [{"_score": "desc"}, {"term_name.keyword": "asc"}],
        "track_total_hits": True,
    }

def normalize_glossary_hit(source):
    return {
        "term_id": source.get("term_id"),
        "term_name": source.get("term_name"),
        "difficulty": source.get("difficulty"),
        "description": source.get("description"),
        "example_text": source.get("example_text"),
        "category": {
            "category_id": source.get("category", {}).get("category_id"),
            "category_name": source.get("category", {}).get("category_name"),
        }
    }

def positive_int(value, default):
    try:
        number = int(value)
    except (TypeError, ValueError):
        return default
    return number if number > 0 else default
